ml-agents-envs/mlagents_envs/rpc_utils.py

```diff
@@ -285,10 +285,22 @@ def steps_from_proto(
     ]
     decision_obs_list: List[np.ndarray] = []
     terminal_obs_list: List[np.ndarray] = []
-    for obs_index, observation_specs in enumerate(behavior_spec.observation_specs):
-        is_visual = len(observation_specs.shape) == 3
+    for obs_index, observation_spec in enumerate(behavior_spec.observation_specs):
```

Comment:
If you're concerned about the overhead of this, I could put it in an `except` following the np.array formation. That would serve as basically cleaning up the error message instead of proactively raising one. Also means we'd want to change _process_maybe_compressed_observation
 and _process_rank_one_or_two_observation to accept an ObservationSpec instead of a tuple (which isn't a bad change either).

Reply:
I thought this error would be caught in C# if an Agent has the wrong dimensions for observations.
Yes, the overhead is a bit concerning, the `except` sounds like a better idea, but if this is not slower than the current code by more than a few percents, it is maybe good enough.

Reply:
The C# side currently "enforces" this via asserts, which won't have any effect in release builds, so I don't think we can rely on that.
I updated this so that the new code (with nice error) only runs in the except block, so there should be no extra overhead (aside from the try)